src/useradd.c

```diff
@@ -1328,15 +1328,22 @@ static void process_flags (int argc, char **argv)
 				if (   ( !VALID (optarg) )
 				    || (   ('\0' != optarg[0])
 				        && ('/'  != optarg[0])
-				        && ('*'  != optarg[0]) )
-				    || (stat(optarg, &st) != 0)
-				    || (S_ISDIR(st.st_mode))
-				    || (access(optarg, X_OK) != 0)) {
+				        && ('*'  != optarg[0]) )) {
 					fprintf (stderr,
 					         _("%s: invalid shell '%s'\n"),
 					         Prog, optarg);
 					exit (E_BAD_ARG);
 				}
+				if (    '\0' != optarg[0]
+				     && '*'  != optarg[0]
+				     && strcmp(optarg, "/sbin/nologin") != 0
+				     && (   stat(optarg, &st) != 0
+				         || S_ISDIR(st.st_mode)
```

Comment:
But if it's a directory then it is not likely to be fixed up later, right? :)

There are quite a few duplicated checks here which I'm worried will eventually (as code shifts) not be handled right.  Would it be better to have a single outer 'if' and then only exit (and maybe change the message) using a nested check for non-executable or missing shell?

Reply:
Perhaps it could be fixed later. Why not? :D The point is, there is no big reason to hard disallow it. I can try to reorg the conditions a little bit.

Reply:
Thinking about it more - I really have no idea how to reorg it in a way that would not be even more convoluted than the existing patch. The problem is that in the second if we just want to avoid trying to stat() empty or a '*...' shell as these are special values and the warning should not be provided for them. However just putting these as higher level if does not work because there is also the !VALID() check which in case of *.... shell should take a precedence.

Reply:
Ok.